qualang_tools/config/gui.py

```diff
@@ -0,0 +1,124 @@
+from threading import Timer
+import argparse
+
+from waitress import serve
+
+
+@app.callback(Output("page-content", "children"), [Input("url", "pathname")])
+def render_page_content(pathname):
+    pathname = str(pathname)
+
+    if pathname == "/":
+        return upload
+    elif pathname.startswith("/config"):
+        return config_editor(pathname)
+    elif pathname == "/download":
+        return download_page
+    # If the user tries to reach a different page, return a 404 message
+    return html.H1("404: Not found", className="text-danger")
+
+
+SIDEBAR_STYLE = {
+    "position": "fixed",
+    "top": 0,
+    "left": 0,
+    "bottom": 0,
+    "width": "16rem",
+    "padding": "2rem 1rem",
+    "background-color": "#f8f9fa",
+}
+
+sidebar = html.Div(
+    [
+        html.Center(html.H3("ConfigBuilder")),
+        html.Hr(),
+        dbc.Nav(
+            [
+                dbc.NavLink(
+                    [html.I(className="bi bi-upload me-2"), "Upload or Start new"],
+                    href="/",
+                    active="exact",
+                ),
+                dbc.NavLink(
+                    [html.I(className="bi bi-sliders me-2"), "View & Edit "],
+                    href="/config",
+                    active="exact",
+                ),
+                dbc.NavLink(
+                    [html.I(className="bi bi-download me-2"), "Download final"],
+                    href="/download",
+                    active="exact",
+                ),
+            ],
+            vertical=True,
+            pills=True,
+        ),
+    ],
+    style=SIDEBAR_STYLE,
+)
+
+CONTENT_STYLE = {
+    "margin-left": "18rem",
+    "margin-right": "2rem",
+    "padding": "2rem 1rem",
+}
+
+content = html.Div(id="page-content", style=CONTENT_STYLE)
+
+app.layout = html.Div([dcc.Location(id="url"), sidebar, content])
+guiserver = app.server
+
+
+def open_browser():
+    global port_number
+    webbrowser.open_new(f"http://localhost:{port_number}")
+
+
+if __name__ == "__main__":
+    global port_number
+
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "-p", "--port", help="port number to use for GUI server", type=int
+    )
+    args = parser.parse_args()
+    if args.port is not None:
+        port_number = args.port
+    else:
+        port_number = 8051
+    host = "127.0.0.1"
+
+    if not os.path.exists(UPLOAD_DIRECTORY):
+        os.makedirs(UPLOAD_DIRECTORY)
+
+    timer = Timer(3, open_browser)
+    timer.start()
+    try:
+        print(f"\tConfig GUI is running on http://{host}:{port_number}")
+        print(
+            "\tIf web browser does not open, please enter the address stated above manually."
+        )
+        print("\tStop GUI by pressing Ctrl+C")
+        serve(app.server, host=host, port=port_number)
+        # for development:
+        # app.run_server(host="127.0.0.1", debug=False, port=port_number)
+    except OSError as e:
+        timer.cancel()
+        if e.args[0] != 98:
```

Comment:
what is this 98?

Reply:
That is error code for "address already in use" when server tries to open port on which there is already process listening. I catching just that specific case.

Reply:
is this error code part of an enum, or you can get the number from somewhere?
if not, please add the explanation as a comment

Reply:
Added comment b9c70e7a04bc29e246f5bdaba53bc0bc69df715f